### PR TITLE
Handling issue #52

### PR DESCRIPTION
**Pull request by amol**
Many consumers of the compute service uses 'OS-EXT-IPS:type' field,
which is easy to support as SL always assigns private
(primaryBackendIpAddress) and public (primaryIpAddress) IPs which can be
categorized into 'fixed' and 'public' respectively.

### CODE REVIEW

**Comment by bodenr** [comment]
IMHO it would be ideal to use the `OS-EXT-IPS:type` to differentiate between a SL static IP and a portable IP -- obviously the later being allocated via a portable IP block. If this were the case the following mapping may be considered:

```
'OS-EXT-IPS:type': 'fixed' --> A SoftLayer static IP
'OS-EXT-IPS:type': 'floating' --> A SoftLayer portable IP (via portable IP block)
```

However I have not done any research into if such details are readily available in the existing SoftLayer python client.

**Comment by sudorandom** [comment]
I agree with Boden here. Both of the primary public and private ips are static.

Even though this isn't an existing directive, but maybe "OS-EXT-IPS:network" would be more appropriate? I want to keep the semantics of OpenStack as similar as possible to the real deal. This might be a case where we can invent a fake extension which reveals what network the IP belongs to. There would be no initial support for clients (or nova itself) but Jumpgate's version would return all the data that you want. The issue that you bring up seems like something that OpenStack users would run into as well.

**Comment by sudorandom** [comment]
Now that I look into this a bit deeper, it does distinguish between public and private networks in the document structure, so I'm a bit confused by the initial issue.

If you look at the `nova list` command, it shows the public ip (if it exists) and the private ip for each instance and is able to determine the difference between them.

In the example below you can see that the public and private IP lists are separated:

```
{
    "servers": [
        {
            "OS-EXT-AZ:availability_zone": 138124,
            "OS-EXT-STS:power_state": 1,
            "OS-EXT-STS:task_state": null,
            "OS-EXT-STS:vm_state": "ACTIVE",
            "accessIPv4": "",
            "accessIPv6": "",
            "addresses": {
                "private": [
                    {
                        "addr": "10.80.12.34",
                        "version": 4
                    }
                ],
                "public": [
                    {
                        "addr": "173.193.12.34",
                        "version": 4
                    }
                ]
            },
            "created": "2012-07-09T13:44:12-07:00",
            "flavor": {
                "id": "1",
                "links": [
                    {
                        "href": "http://localhost:5000/compute/v2/flavors/1",
                        "rel": "bookmark"
                    }
                ]
            },
            "hostId": 12345,
            "id": "12345",
            "image_name": "",
            "links": [
                {
                    "href": "http://localhost:5000/compute/v2/1234/servers/12345",
                    "rel": "self"
                }
            ],
            "name": "example",
            "security_groups": [
                {
                    "name": "default"
                }
            ],
            "status": "ACTIVE",
            "tenant_id": 1234,
            "updated": "2013-05-29T17:32:19-07:00"
        }
    ]
}
```

http://api.openstack.org/api-ref-compute.html#compute_server-addresses

**Comment by amol** [comment]
This fix has been introduced due some service such new one "Savanna" service [1] needs OS-EXT-IPS field to distinguish. Agree with Boden's comment.

[1] https://github.com/openstack/savanna/blob/980db44bdadae00a453c24e55c289dce5f044791/savanna/service/networks.py#L46

**Comment by bodenr** [comment]
@amol -- maybe there is some confusion here... My comment above was about using the SoftLayer API to determine if the IP address returned from the CCI is a static IP address or portable IP address and then using `floating` for portable and `fixed` for static... Based on your changes it seems you are assuming public is always fixed and private is always floating, but I do not believe that is the case... I think we'd need more details from the SLAPI to determine per IP address.

File: jumpgate/compute/drivers/sl/servers.py [hunk 1]
```diff
@@ -402,12 +402,14 @@ def get_server_details_dict(app, req, instance):
...
             'addr': instance.get('primaryBackendIpAddress'),
             'version': 4,
+            'OS-EXT-IPS:type': 'fixed',
         }]
 
     if instance.get('primaryIpAddress'):
         addresses['public'] = [{
             'addr': instance.get('primaryIpAddress'),
             'version': 4,
+            'OS-EXT-IPS:type': 'fixed',
```
Comment (amol) [inline jumpgate/compute/drivers/sl/servers.py:412]: Converted to `fixed`. There is no direct way to find portable IP from `CCIManager`, skipping adding floating IP to the details of the instance.

**Comment by sudorandom** [comment]
+1 LGTM

**Pull request merged**